compiler/src/dotty/tools/backend/jvm/GenBCode.scala

```diff
@@ -190,19 +190,23 @@ class GenBCodePipeline(val entryPoints: List[Symbol], val int: DottyBackendInter
         val claszSymbol = cd.symbol
 
         // GenASM checks this before classfiles are emitted, https://github.com/scala/scala/commit/e4d1d930693ac75d8eb64c2c3c69f2fc22bec739
-        val lowercaseJavaClassName = claszSymbol.name.toString.toLowerCase
-        caseInsensitively.get(lowercaseJavaClassName) match {
-          case None =>
-            caseInsensitively.put(lowercaseJavaClassName, claszSymbol)
-          case Some(dupClassSym) =>
-            // Order is not deterministic so we enforce lexicographic order between the duplicates for error-reporting
-            if (claszSymbol.name.toString < dupClassSym.name.toString)
-              ctx.warning(s"Class ${claszSymbol.name} differs only in case from ${dupClassSym.name}. " +
-                          "Such classes will overwrite one another on case-insensitive filesystems.", claszSymbol.pos)
-            else
-              ctx.warning(s"Class ${dupClassSym.name} differs only in case from ${claszSymbol.name}. " +
-                          "Such classes will overwrite one another on case-insensitive filesystems.", dupClassSym.pos)
+        def checkName(claszSymbol: Symbol): Unit = {
+          val lowercaseJavaClassName = claszSymbol.name.toString.toLowerCase
+          caseInsensitively.get(lowercaseJavaClassName) match {
+            case None =>
+              caseInsensitively.put(lowercaseJavaClassName, claszSymbol)
+            case Some(dupClassSym) =>
+              // Order is not deterministic so we enforce lexicographic order between the duplicates for error-reporting
+              val (cl1, cl2) =
+                if (claszSymbol.name.toString < dupClassSym.name.toString) (claszSymbol, dupClassSym)
+                else (dupClassSym, claszSymbol)
+              ctx.warning(s"Class ${cl1.name} differs only in case from ${cl2.name}. " +
+                          "Such classes will overwrite one another on case-insensitive filesystems.", cl1.pos)
+          }
         }
+        checkName(claszSymbol)
+        if (int.symHelper(claszSymbol).isTopLevelModuleClass)
```

Comment:
I don't think this should be limited to top-level modules, the same problem exists for inner modules no?